lib/facebook.ex

```diff
@@ -616,6 +712,32 @@ defmodule Facebook do
     |> ResponseFormatter.format_response()
   end
 
+  @doc """
+  Decodes a signed request from a client SDK (in-app payments), verifies the
+  signature and -if valid- returns its contents.
+  """
+  @spec decode_signed_request(signed_request) :: resp
+  def decode_signed_request(signed_request) when is_binary(signed_request) do
+    with [signature_str | [payload_str | _]] when
+           is_binary(signature_str) and is_binary(payload_str)
+           <- String.split(signed_request, "."),
+         {:ok, signature} <- Base.url_decode64(signature_str),
+         _signature_verification = ^signature <- signature(payload_str),
+         {:ok, payload} <- Base.url_decode64(payload_str),
+         {:ok, payload} <- JSON.decode(payload)
```

Comment:
IMO this should be doable in a simpler way. 
What's the purpose of the guard in the function? If `signed_request` is not a binary and you call that function then there won't be another function executed instead. 
Also why the `with` with another guard (if you split a binary it should still be a binary)?

Reply:
I'm quite new to Elixir, maybe it really can be done better: I used the inner guards to ensure there won't be nil values. Which, you're right, is complete nonsense in the case of split() as it doesn't return nil values. I'll remove them.
However, guard'ing the function itself prevents it from being called with a wrong value and raise an exception instead of trying to call split() with a wrong argument. What do you think about that? How could I do it better?